### DISCUSSION TITLE
Signals has no drivers when created the AST in a test case

### DISCUSSION THREAD

**Discussion by Sustrak** [MikePopoloski/slang#713] [Q&A] [answered]
Hi,
the problem is the following. I have this test:

```c++
TEST_CASE("Not assigned on reset") {
    auto tree = SyntaxTree::fromText(R"(
module top;
    logic clk;
    logic rst_n;
    logic a, b;

    always_ff @(posedge clk or negedge rst_n) begin
        if (~rst_n) begin
            a <= '0;
            b <= 1'b1;
        end else begin
            a <= 1'b1;
        end
    end
endmodule
)");

    Compilation compilation;
    compilation.addSyntaxTree(tree);
    auto& root = compilation.getRoot();

    OnlyAssignedOnReset visitor;
    bool result = visitor.check(root);
    CHECK(result == false);
}
```

Late on, in the code inside `OnlyAssignedOnReset` :
```c++
void handle(const VariableSymbol& symbol) {
        if (symbol.drivers().empty())
            return;

        auto firstDriver = *symbol.drivers().begin();
...
```

When variables `a` and `b` are visited `symbol.drivers().empty()` returns `true`. When they are driven inside the `always_ff` block.

If instead of creating a test to run the `OnlyAssignedOnReset` I do it through the `driver` the visitor works as intended since `symbol.drivers().empty()`  is not empty.

Do you have any idea why I'm seeing this difference in the execution?

Thanks

**Comment by MikePopoloski** [comment, marked as answer]
AST nodes are lazily created and evaluated, so you need to visit the whole tree in order to ensure you have found all of the drivers. The getAllDiagnostics function will do that for you (internally using an ElabVisitor to do it) so you can just make sure that's called before doing any of your own checks.

**Reply by Sustrak** [reply]
That did the trick. Thanks :smile:
